Core/include/Acts/Utilities/KDTree.hpp

```diff
@@ -318,369 +310,249 @@ class KDTree {
+        std::size_t lhs_size = std::distance(this->m_begin_it, pivot);
+        std::size_t rhs_size = std::distance(pivot, this->m_end_it);
+
+        // Next, we check whether the left-hand node should be another internal
+        // node or a leaf node, and we construct the node recursively.
+        this->m_lhs = std::make_unique<KDTreeNode>(
+            this->m_begin_it, pivot,
+            lhs_size > LeafSize ? NodeType::INTERNAL : NodeType::LEAF);
+
+        // Same on the right hand side.
+        this->m_rhs = std::make_unique<KDTreeNode>(
+            pivot, this->m_end_it,
+            rhs_size > LeafSize ? NodeType::INTERNAL : NodeType::LEAF);
       }
     }
 
-    /// @brief Debugging string method for this (sub-)k-d tree.
+    /// @brief Perform a range search in the k-d tree, mapping the key-value
+    /// pairs to a side-effecting function.
     ///
-    /// This prints information to stdout about the structure of the (sub-)tree
-    /// defined by this node. Not designed for use in real code.
+    /// This is the most powerful range search method we have, assuming that we
+    /// can use arbitrary side effects, which we can. All other range search
+    /// methods are implemented in terms of this particular function.
     ///
-    /// @param i The amount of indentation to use.
-    std::string toString(std::size_t i) const override {
-      std::stringstream out;
+    /// @param r The range to search for.
+    /// @param f The mapping function to apply to matching elements.
+    template <typename Callable>
+    void rangeSearchMapDiscard(const range_t &r, Callable &&f) const {
+      // Determine whether the range completely covers the bounding box of
+      // this leaf node. If it is, we can copy all values without having to
+      // check for them being inside the range again.
+      bool contained = r >= this->m_range;
 
-      // First, we print some indentation to make the output look nicer.
-      for (std::size_t j = 0; j < i; ++j) {
-        out << " ";
-      }
+      if (m_type == NodeType::INTERNAL) {
+        // Firstly, we can check if the range completely contains the bounding
+        // box of this node. If that is the case, we know for certain that any
+        // value contained below this node should end up in the output, and we
+        // can stop recursively looking for them.
+        if (contained) {
+          // We can also pre-allocate space for the number of elements, since we
+          // are inserting all of them anyway.
+          for (iterator_t i = this->m_begin_it; i != this->m_end_it; ++i) {
+            f(i->first, i->second);
+          }
+
+          return;
+        }
 
-      // Print some information about the node.
-      out << "Node with range " << this->m_range[m_dim].min() << " to "
-          << this->m_mid << " to " << this->m_range[m_dim].max()
-          << " (total size " << this->size() << ")" << std::endl;
+        assert(this->m_lhs && this->m_rhs);
```

Comment:
Consider adding a message like
```suggestion
        assert(this->m_lhs && this->m_rhs && "Did not find lhs and rhs");
```

Reply:
Done.